### PR TITLE
ENH - add dual annealing

### PR DESCRIPTION
**Pull request by sgubianpm**
Dear @antonior92, @andyfaff, @js850, @rgommers and all,

Thanks a lot for your time spent on the review. I spent some extra time addressing the latest comments as the next release is approaching. For reference, the old PRs were:
#6569
#6197
#8203
Here comes a new PR as the previous ones were very long and old.
This one comes with the main following features:
- Improved doc string as suggested in the comments
- Better test coverage with additional unit tests as commented in #8203 
- Better separation of concerns as proposed in the comments
- Callback mechanism implemented
- gaussian function replaced with the SciPy one
- previous PR's comments addressed
- better naming for the optimization function (short and describing the method)

Many thanks!
Best regards,

Sylvain.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks for the update @sgubianpm. I'll be offline for a couple of weeks, but marked it for the next release (1.2.0) as a reminder to come back to it.

File: scipy/optimize/_dual_annealing.py [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+            else:
+                init_error = False
+            # If first time reset, initialize ebest and xbest
+            if self.ebest is None and self.xbest is None:
+                self.ebest = self.current_energy
+                self.xbest = np.copy(self.current_location)
+            # Otherwise, we keep them in case of reannealing reset
+
+
+class MarkovChain(object):
```
Comment (js850) [inline scipy/optimize/_dual_annealing.py]: This class is doing more than just a MarkovChain.  The single responsibility principle would argue that, if possible you should split out the part that's doing a single MarkovChain from the rest of what this class is doing.  If that is not possible or appropriate, then you should probably rename this class

File: scipy/optimize/_dual_annealing.py [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+                    self.not_improved_idx = 0
+            else:
+                # We have not improved but do we accept the new location?
+                self.accept_reject(j, e, x_visit)
+            if self.func_wrapper.nfev >= self.func_wrapper.maxfun:
+                return False
+        # End of MarkovChain loop
+        return True
+
+    def track_better_minimum(self, e, x, context):
```
Comment (js850) [inline scipy/optimize/_dual_annealing.py]: Suggestion: I think this function fits better as part of EnergyState.  Then you could wrap this functionality into an update method similar to basinhopping.Storage.update().  Then EnergyState has all the responsibility for keeping track of the lowest state and MonteCarlo doesn't have to worry about it anymore.

**Review by js850 — COMMENTED** [review]
I just had a few suggestions, but in general it's looking good 👍.

File: scipy/optimize/_dual_annealing.py (hunk 2 above)
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Agreed. Moved to EnergyState class.

File: scipy/optimize/_dual_annealing.py [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+        # Use the new point only if it is valid and return a better results
+        if is_valid and mres.fun < e:
+            return mres.fun, mres.x
+        else:
+            return e, x_tmp
+
+
+def dual_annealing(func, x0, bounds, args=(), maxiter=1000,
+                   local_search_options={}, initial_temp=5230., visit=2.62,
+                   accept=-5.0, maxfun=1e7, seed=None, no_local_search=False,
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: I agree. I also thougth it was misleading but I think that users are used to the other scipy interfaces, reason why I kept it like that.

File: scipy/optimize/_dual_annealing.py [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+    minimizer_wrapper = LocalSearchWrapper(
+        bounds, func_wrapper, **local_search_options)
+    # Initialization of RandomState for reproducible runs if seed provided
+    rand_state = check_random_state(seed)
+    # Initialization of the energy state
+    energy_state = EnergyState(lower, upper)
+    energy_state.reset(func_wrapper, rand_state, x0)
+    # Minimum value of annealing temperature reached to perform
+    # re-annealing
+    temperature_restart = 0.1
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: New parameter restart_temp_ratio exposed to address this. Thanks!

File: scipy/optimize/_dual_annealing.py [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+                energy_state.reset(func_wrapper, rand_state)
+                break
+            # starting Markov chain
+            val = markov_chain.run(i, temperature)
+            if not val:
+                if func_wrapper.nfev >= maxfun:
+                    message = ["Maximum number of function calls reached"]
+                else:
+                    message = ["callback function requested stop early by "
+                               "returning True"]
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Added methods update_current and update_best (which also trigger the callback if defined) in EnergyState to address that. Thanks!

File: scipy/optimize/_dual_annealing.py [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            if not j or self.energy_state.current_energy < self.emin:
+                self.emin = self.energy_state.current_energy
+                self.xmin = np.copy(self.energy_state.current_location)
+
+    def run(self, step, temperature):
+        self.temperature_step = temperature / float(step + 1)
+        self.not_improved_idx += 1
+        for j in range(self.energy_state.current_location.size * 2):
+            if not j:
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Changed to `j == 0`

File: scipy/optimize/_dual_annealing.py (hunk 1 above)
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Classed renamed into StrategyChain which reflects better the purpose.

File: scipy/optimize/_dual_annealing.py (hunk 2 above)
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Added methods update_current and update_best (which also trigger the callback if defined) in EnergyState to address that. Thanks!

File: scipy/optimize/_dual_annealing.py [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+class ObjectiveFunWrapper(object):
+
+    def __init__(self, func, maxfun=1e7, *args):
+        self.func = func
+        self.args = args
+        self.nfev = 0
+        self.maxfun = maxfun
+
+    def fun(self, x):
+        self.nfev += 1
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Added test and counting number of call to gradient.

**Comment by sgubianpm** [comment]
Dear @js850 , @antonior92 , @rgommers , @andyfaff 
Thanks all for your time. Anyt other comments/suggestions? (as the release date is approaching).
The current travis-ci is failling due to `newton` module in `scipy.optimize` that is not convergin after 50 iterations accoring the log (for 2 cases).

**Comment by rgommers** [comment]
gh-8203 has benchmarks, but I don't see the code to reproduce them. I'd like to run the benchmarks for the latest version of this code. could you point me to your benchmark code @sgubianpm

**Comment by sgubianpm** [comment]
@rgommers,thanks for the note. The extensive benchmark suite is available [here](https://github.com/sgubianpm/GOBench).
Running the full benchmark (261 testing functions on all the methods for 100 runs) is very CPU demanding and took a week using 272 cores of the cluster. However, but playing with the options of the benchmark suite, you can do some lighter/faster benchmarks using only few functions of your choice, few methods and limit the number of runs (default is 100). There is also the `multidim` option that enables the multi dimension benchmark (use some testing functions that have a generalized analytic form to n dimension to be tested from dim 5 to 100). Enabling this one show how `DA` performs well with a big number of parameters.
I hope this helps.

**Comment by sgubianpm** [comment]
@rgommers, @js850, @antonior92, @andyfaff, the current version of the code has been benchmarked and gives the following results:
go_benchmark-1e-6
The corresponding csv file generated can be retrieved:
[results_20181026.xlsx](https://github.com/scipy/scipy/files/2519234/results_20181026.xlsx)

As mentioned in the previous comment, the benchmarking code is available [here](https://github.com/sgubianpm/GOBench).

**Comment by js850** [comment]
@sgubianpm , I suspect the strong showing of dual annealing is strongly depend on the fact that you set the default local optimizer to be L-BFGS-B and have tuned the parameters, e.g. `LS_MAXITER_MIN = 100`.  In contrast, BasinHopping uses the default optimizer BFGS with default parameters.  LBFGS is far superior to this BFGS implementation for many function.  Can you please re-run the benchmarks (or a subset of them) when they both use the same local optimizer?

**Comment by rgommers** [comment]
Thanks for the benchmark results @sgubianpm. Would be interested in the answer to @js850's question, but either way the performance looks definitely good enough!

CI failures look unrelated.

1.2.x branch date is 5 Nov, so let's aim to merge this before then.

**Comment by sgubianpm** [comment]
@js850, @rgommers, thanks a lot for your comments. In GOBench, the results for `BH` shown in the previous comments were already using L-BFGS-B rather than BFGS because bounds were provided so that `minimize` function automatically switch BFGS to L-BFGS-B. ([code here](https://github.com/sgubianpm/GOBench/blob/bbe61b4b0fbcd79b5c1cb5551cd09334b0accea1/gobench/bench.py#L284-L301)).

As proposed by @js850, new entries have been added in the benchmark  to compare even more apples to apples:
- BH-MI (Basinhopping with maxiter of the local minimizer set the same way as DA)
[Code here](https://github.com/sgubianpm/GOBench/blob/1cf85f2f44ba89679533db5789a3ffa1b7bbfb1a/gobench/bench.py#L303-L328)
- BH-BFGS (Basinhopping forced to use BFGS (and therefore no bounds provided to the LS), which is BH default settings)
[Code here](https://github.com/sgubianpm/GOBench/blob/1cf85f2f44ba89679533db5789a3ffa1b7bbfb1a/gobench/bench.py#L331-L345)
- DA-BFGS (Dual annealing forced to use BFGS  rather than L-BFGS-B, exact same settings as BH-BFGS)
[Code here](https://github.com/sgubianpm/GOBench/blob/1cf85f2f44ba89679533db5789a3ffa1b7bbfb1a/gobench/bench.py#L184-L198)

The results plot is provided here:
go_benchmark-1e-6-with-added-tests

The generated csv file is provided here:
[results_20181029_2.xlsx](https://github.com/scipy/scipy/files/2525938/results_20181029_2.xlsx)

Summary:
* For `DA`, using BFGS or L-BFGS-B does not significantly affect the results of `DA`
* For `BH`, using L-BFGS-B or BFGS does not significantly affect the successful rate of `BH` but using L-BFGS-B does improve the average number of objective function calls of `BH` compared to BFGS, especially when the same strategy of `maxiter` for the local search is applied as in `DA`.


Any other suggestions/comments? Don't want to miss the release this time.

**Comment by rgommers** [comment]
I just pushed a commit to fix up some docstring formatting issues. Documentation looks good to go now.

I'll have a look at the benchmark as well, just checking it runs (won't do the 100 repeats ...). I think we should add `dual_annealing` to the benchmarks in this repo, but I'm fine with merging this PR first.

**Comment by rgommers** [comment]
Benchmarks work like a charm and I can reproduce the results in the README. So all good to go. Appveyor is backed up, but let's not wait for that - in it goes.

Thanks @sgubianpm for the persistence and new optimizer with really good performance!

**Comment by rgommers** [comment]
@sgubianpm would you mind editing https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0#scipyoptimize-improvements with a couple of sentences about `dual_annealing`?

**Comment by rgommers** [comment]
I opened gh-9415 to keep track of the todo item of adding to the scipy benchmarks.

**Review by js850 — COMMENTED** [review]
Thanks for the more detailed benchmarks!  I left a few minor comments, but in general I think it's in good shape and could be ready to merge.

File: scipy/optimize/_dual_annealing.py [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+            self.kwargs['options'] = {
+                'maxiter': ls_max_iter,
+            }
+            self.kwargs['bounds'] = list(zip(self.lower, self.upper))
+
+    def local_search(self, x, e):
+        # Run local search from the given x location where energy value is e
+        x_tmp = np.copy(x)
+        mres = self.minimizer(self.func_wrapper.fun, x, **self.kwargs)
+        if 'njev' in mres.keys():
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Thanks @rgommers for the documentation fixes!
Thanks @js850, I will create a new PR for addressing these points. Maybe I need to add also the ``versionadded instruction`` in the documentation in the code (as explain in the PR checklist).

[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Don't know if the `njev` is used either but there is no problem for me to add it.

[the remark this answers is not in the document]
Comment (js850) [reply scipy/optimize/_dual_annealing.py]: Haha, I didn't notice it had already been merged :)

File: scipy/optimize/_dual_annealing.py [hunk 9]
```diff
@@ -0,0 +1,666 @@
...
+            e - self.energy_state.current_energy) / (
+                self.temperature_step + 1.)
+        if pqv_temp <= 0.:
+            pqv = 0.
+        else:
+            pqv = np.exp(np.log(pqv_temp) / (
+                1. - self.acceptance_param))
+        if r <= pqv:
+            # We accept the new location and update state
+            self.energy_state.update_current(e, x_visit)
```
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: Thanks. The idea here to say, ok this point is accepted but it is not yet the best location. So calling update_current was ok.

File: scipy/optimize/_dual_annealing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py]: @js850 , Thanks. I addressed it in PR #9424

File: scipy/optimize/_dual_annealing.py [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+        The value range is (0, 3].
+    rand_state : RandomState object
+        A numpy.random.RandomState object for using the current state of the
+        created random generator container.
+    """
+    TAIL_LIMIT = 1.e8
+    MIN_VISIT_BOUND = 1.e-10
+
+    def __init__(self, lb, ub, visiting_param, rand_state):
+        self.visiting_param = visiting_param
```
Comment (ahojukka5) [inline scipy/optimize/_dual_annealing.py:50]: An exception should be thrown for visiting parameter out of range.

File: scipy/optimize/_dual_annealing.py [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+        A 1-D numpy ndarray containing lower bounds of the generated
+        components. Neither NaN or inf are allowed.
+    ub : array_like
+        A 1-D numpy ndarray containing upper bounds for the generated
+        components. Neither NaN or inf are allowed.
+    visiting_param : float
+        Parameter for visiting distribution. Default value is 2.62.
+        Higher values give the visiting distribution a heavier tail, this
+        makes the algorithm jump to a more distant region.
+        The value range is (0, 3].
```
Comment (ahojukka5) [inline scipy/optimize/_dual_annealing.py:41]: It looks that the range is not (0, 3]. 1.4 and under gives nans to x.

File: scipy/optimize/_dual_annealing.py [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+    def visit_fn(self, temperature):
+        """ Formula Visita from p. 405 of reference [2] """
+        factor1 = np.exp(np.log(temperature) / (self.visiting_param - 1.0))
+        factor2 = np.exp((4.0 - self.visiting_param) * np.log(
+            self.visiting_param - 1.0))
+        factor3 = np.exp((2.0 - self.visiting_param) * np.log(2.0) / (
+            self.visiting_param - 1.0))
+        factor4 = np.sqrt(np.pi) * factor1 * factor2 / (factor3 * (
+            3.0 - self.visiting_param))
+        factor5 = 1.0 / (self.visiting_param - 1.0) - 0.5
```
Comment (ahojukka5) [inline scipy/optimize/_dual_annealing.py:105]: ZeroDivisionError: float division by zero for visiting parameter 1.0

**Pull request merged**